openedx/core/lib/block_cache/block_structure.py

```diff
@@ -0,0 +1,588 @@
+                # Add it to the new pruned structure
+                self._add_block(pruned_block_relations, block_key)
+
+                # Add a relationship to only those old children that
+                # were also added to the new pruned structure.
+                for child in old_block_relations[block_key].children:
+                    if child in pruned_block_relations:
+                        self._add_to_relations(pruned_block_relations, block_key, child)
+
+        # Replace this structure's relations with the newly pruned one.
+        self._block_relations = pruned_block_relations
+
+    def _get_block_keys(self):
+        """
+        Returns an iterator of all the block keys in the block
+        structure.
+        """
+        return self._block_relations.iterkeys()
+
+    def _add_relation(self, parent_key, child_key):
+        """
+        Adds a parent to child relationship in this block structure.
+
+        Arguments:
+            parent_key (UsageKey) - Usage key of the parent block.
+            child_key (UsageKey) - Usage key of the child block.
+        """
+        self._add_to_relations(self._block_relations, parent_key, child_key)
+
+    @staticmethod
+    def _add_to_relations(block_relations, parent_key, child_key):
+        """
+        Adds a parent to child relationship in the given block
+        relations map.
+
+        Arguments:
+            block_relations (defaultdict({UsageKey: _BlockRelations})) -
+                Internal map of a block's usage key to its
+                parents/children relations.
+
+            parent_key (UsageKey) - Usage key of the parent block.
+
+            child_key (UsageKey) - Usage key of the child block.
+        """
+        block_relations[child_key].parents.append(parent_key)
+        block_relations[parent_key].children.append(child_key)
+
+    @staticmethod
+    def _add_block(block_relations, usage_key):
+        """
+        Adds the given usage_key to the given block_relations map.
+
+        Arguments:
+            block_relations (defaultdict({UsageKey: _BlockRelations})) -
+                Internal map of a block's usage key to its
+                parents/children relations.
+
+            usage_key (UsageKey) - Usage key of the block that is to
+                be added to the given block_relations.
+        """
+
+        # Since block_relations is a DefaultDict, the usage_key will
+        # automatically be added by virtue of accessing it.
+        _ = block_relations[usage_key]
+
+
+class _BlockData(object):
+    """
+    Data structure to encapsulate collected data for a single block.
+    """
+    def __init__(self):
+        # Map of xblock field name to the field's value for this block.
+        # dict {string: any picklable type}
+        self._xblock_fields = {}
+
+        # Map of transformer name to the transformer's data for this
+        # block.
+        # defaultdict {string: dict}
+        self._transformer_data = defaultdict(dict)
```

Comment:
The leading underscores is something that was added later to denote attributes and methods that are specifically for internal usage within the block cache layer.  So this was a code-wide change.  I agree that the attributes declared within an internal class don't necessarily need to be declared internal.